csharp/ql/src/semmle/code/csharp/Namespace.qll

```diff
@@ -112,8 +112,12 @@ class Namespace extends DotNet::Namespace, TypeContainer, @namespace {
 
   override predicate fromLibrary() { not this.fromSource() }
 
-  /** Gets the URL of this namespace (which is empty in this case). */
-  string getURL() { result = "" }
+  /** Gets a declaration of this namespace, if any. */
+  NamespaceDeclaration getADeclaration() { result.getNamespace() = this }
+
+  override Location getALocation() {
+    result = this.getADeclaration().getALocation()
```

Comment:
What happens for namespaces that don't have a declaration? These could happen for namespaces provided only in assemblies for example.

This could confuse results quite a bit in the case that there are multiple declarations in source code. In that case, `getLocation()` also returns multiple results. Luckily, I don't think we have any queries that return namespaces.

Perhaps we should return a dummy location here?

Reply:
Namespaces with no declarations will get a `file://0:0:0:0` URL, and I think that should be fine.

I realize that `getLocation()` for namespaces defined in source code can now yield multiple results, but I personally think that is fine, just like `getLocation()` will return all locations for partial classes.

Reply:
I was just checking the behaviour in Eclipse, and it seems that `file://:0:0:0:0` won't get rendered, and instead the result would get hidden. That could be quite confusing for users. Instead, I would advocate just removing `getURL` and `getLocation` completely from `Namespace`, as this will still render the result, but with no location which I think is fine and avoids other problems.

Reply:
OK, in that case I need to reintroduce `getURL()`, but only for entities that may not have a location.

Reply:
So, I have reintroduced `getURL()`, but this time as a cached predicate.